kuma/users/tests/test_auth.py

```diff
@@ -0,0 +1,107 @@
+import pytest
+
+from django.test import RequestFactory
+
+from kuma.users.auth import (
+    InvalidClaimsError,
+    KumaOIDCAuthenticationBackend,
+    logout_url,
+)
+from kuma.users.models import UserProfile
+
+
+@pytest.mark.django_db
+def test_bark_on_no_email_claims():
+    backend = KumaOIDCAuthenticationBackend()
+    with pytest.raises(InvalidClaimsError):
+        backend.filter_users_by_claims({})
+    # Sanity check that it works when you supply an email
+    qs = backend.filter_users_by_claims({"email": "peterbe@example.com"})
+    assert qs.count() == 0
+
+
+@pytest.mark.django_db
+def test_create_and_update_user_not_subscriber():
+    backend = KumaOIDCAuthenticationBackend()
+    user = backend.create_user(
+        {
+            "email": "peterbe@example.com",
+        }
+    )
+    assert user.email == "peterbe@example.com"
+    user_profile = UserProfile.objects.get(user=user)
+    assert not user_profile.is_subscriber
+    assert not user_profile.subscriber_number
+
+
+@pytest.mark.django_db
+def test_create_and_update_user_new_subscriber():
+    backend = KumaOIDCAuthenticationBackend()
+    user = backend.create_user(
+        {
+            "email": "peterbe@example.com",
+            "subscriptions": ["mdn_plus", "other_vpn_thing"],
+        }
+    )
+    assert user.email == "peterbe@example.com"
+    user_profile = UserProfile.objects.get(user=user)
+    assert user_profile.is_subscriber
+    assert user_profile.subscriber_number == 1
+
+    backend.update_user(
+        user,
+        {
+            "email": "peterbe@example.com",
+            "subscriptions": ["other_vpn_thing"],  # Note!
+        },
+    )
+    user_profile.refresh_from_db()
+    assert not user_profile.is_subscriber
+    assert user_profile.subscriber_number == 1
+
+
+@pytest.mark.django_db
+def test_update_user_from_not_to_new_subscriber():
+    backend = KumaOIDCAuthenticationBackend()
+    user = backend.create_user(
+        {
+            "email": "peterbe@example.com",
+        }
+    )
+    assert user.email == "peterbe@example.com"
+    user_profile = UserProfile.objects.get(user=user)
+    assert not user_profile.is_subscriber
+    assert not user_profile.subscriber_number
+
+    backend.update_user(
+        user,
+        {
+            "email": "peterbe@example.com",
+            "subscriptions": ["mdn_plus"],  # Note!
+        },
+    )
+    user_profile.refresh_from_db()
+    assert user_profile.is_subscriber
+    assert user_profile.subscriber_number == 1
+
```

Comment:
Can you add some tests for the subscriber number? I added the following locally when reviewing, but feel free to adjust as you wish:

```python
@pytest.mark.django_db
def test_new_profile_subscriber_number():
    backend = KumaOIDCAuthenticationBackend()
    user = backend.create_user(
        {
            "email": "peterbe@example.com",
            "subscriptions": ["mdn_plus", "other_vpn_thing"],
        }
    )
    assert user.email == "peterbe@example.com"
    user_profile = UserProfile.objects.get(user=user)
    assert user_profile.is_subscriber
    assert user_profile.subscriber_number == 1

    user = backend.create_user(
        {
            "email": "ringo@beatles.com",
            "subscriptions": ["mdn_plus"],
        }
    )
    assert user.email == "ringo@beatles.com"
    user_profile = UserProfile.objects.get(user=user)
    assert user_profile.is_subscriber
    assert user_profile.subscriber_number == 2


@pytest.mark.django_db
def test_existing_profile_subscriber_number():
    backend = KumaOIDCAuthenticationBackend()
    first_user = backend.create_user(
        {
            "email": "peterbe@example.com",
            "subscriptions": ["other_vpn_thing"],
        }
    )
    assert first_user.email == "peterbe@example.com"
    first_user_profile = UserProfile.objects.get(user=first_user)
    assert not first_user_profile.is_subscriber
    assert not first_user_profile.subscriber_number

    second_user = backend.create_user(
        {
            "email": "ringo@beatles.com",
            "subscriptions": ["mdn_plus"],
        }
    )
    assert second_user.email == "ringo@beatles.com"
    second_user_profile = UserProfile.objects.get(user=second_user)
    assert second_user_profile.is_subscriber
    assert second_user_profile.subscriber_number == 1

    backend.update_user(
        first_user,
        {
            "email": "peterbe@example.com",
            "subscriptions": ["mdn_plus", "other_vpn_thing"],
        },
    )
    first_user_profile.refresh_from_db()
    assert first_user_profile.is_subscriber
    assert first_user_profile.subscriber_number == 2
```

Reply:
I didn't copy that verbatim but I did update my existing tests to...

- make sure the next user becomes subscriber_number=2
- when an existing subscriber (who was subscriber_number=1) ceases subscription, the next (new) subscribing user becomes subscriber_number=2